src/main/java/io/vertx/core/eventbus/impl/clustered/ConnectionHolder.java

```diff
@@ -0,0 +1,130 @@
+package io.vertx.core.eventbus.impl.clustered;
+
+import io.vertx.core.Vertx;
+import io.vertx.core.VertxOptions;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.eventbus.impl.codecs.PingMessageCodec;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.core.net.NetClient;
+import io.vertx.core.net.NetClientOptions;
+import io.vertx.core.net.NetSocket;
+import io.vertx.core.net.impl.NetClientImpl;
+import io.vertx.core.net.impl.ServerID;
+import io.vertx.core.spi.metrics.EventBusMetrics;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+/**
+ * @author <a href="http://tfox.org">Tim Fox</a>
+ */
+class ConnectionHolder {
+
+  private static final Logger log = LoggerFactory.getLogger(ConnectionHolder.class);
+
+  private static final String PING_ADDRESS = "__vertx_ping";
+
+  private final ClusteredEventBus eventBus;
+  private final NetClient client;
+  private final ServerID serverID;
+  private final Vertx vertx;
+  private final EventBusMetrics metrics;
+
+  private Queue<ClusteredMessage> pending;
+  private NetSocket socket;
+  private boolean connected;
+  private long timeoutID = -1;
+  private long pingTimeoutID = -1;
+
+  ConnectionHolder(ClusteredEventBus eventBus, ServerID serverID) {
+    this.eventBus = eventBus;
+    this.serverID = serverID;
+    this.vertx = eventBus.vertx();
+    this.metrics = eventBus.getMetrics();
+    client = new NetClientImpl(eventBus.vertx(), new NetClientOptions().setConnectTimeout(60 * 1000), false);
```

Comment:
Ack. But note that this is a straight refactor - I have not changed the old functionality, so this was an issue before, and this comment applies to most of your comments below. Instead it might be better to open issues for these points you notice as fixing them is not part of the task I am doing right now.